src/storage/tracked.rs

```diff
@@ -0,0 +1,378 @@
+use std::iter::Enumerate;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::slice;
+
+use hibitset::{BitSet, BitSetLike};
+
+use {Component, DenseVecStorage, Index, MaskedStorage, Storage, UnprotectedStorage};
+
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub enum Change {
+    #[doc(hidden)]
+    None,
+    Inserted,
+    Modified,
+    Removed,
+}
+
+impl Change {
+    fn add(&mut self, change: Change) {
+        let new = match (*self, change) {
+            (Change::None, new) => new,
+            (Change::Inserted, Change::Modified) => Change::Inserted,
+            (Change::Inserted, Change::Removed) => Change::None,
+            (Change::Modified, Change::Modified) => Change::Modified,
+            (Change::Modified, Change::Removed) => Change::Removed,
+            (Change::Removed, Change::Inserted) => Change::Modified,
```

Comment:
I would `use` the variants of the enum for this method to make it cleaner, but using them would be confusing for the `None` variant.
````rust
let new = match (*self, change) {
    (None, new) => new,
    (Inserted, Modified) => Inserted,
    (Inserted, Removed) => None,
    (Modified, Modified) => Modified,
    (Modified, Removed) => Removed,
    (Removed, Inserted) => Modified,
    (old, new) => panic!("Didn't expect change from {:?} to {:?}", old, new),
};
````
Also could this use `Add` trait?

Reply:
> Also could this use Add trait?

This has a very different meaning than mathematical addition and I also don't want to expose it.